image/src/routes/metadata.ts

```diff
@@ -0,0 +1,102 @@
+import type { Context } from 'hono'
+import type { Env } from '../utils/constants'
+import { normalize, contentFrom, type BaseMetadata } from '@kodadot1/hyperdata'
+import { getProviderList } from '@kodadot1/minipfs'
+import { ipfsUrl, toNftstorage } from '../utils/ipfs'
+import { encodeEndpoint } from './type-url'
+
+type HonoInterface = Context<
+  {
+    Bindings: Env
+  },
+  '/metadata',
+  {}
+>
+
+const getMimeType = async (url: string) => {
+  const externalUrl = url.includes('kodadot.xyz') ? toNftstorage(url) : url
+  const data = await fetch(externalUrl, { method: 'HEAD' })
+  const contentType = data.headers.get('content-type')
+
+  return contentType
+}
+
+const toCustomKodaURL = (url: string) => {
+  if (!url) {
+    return ''
+  }
+
+  const kodaUrl = new URL(getProviderList(['kodadot_beta'])[0])
+  kodaUrl.pathname = '/type/url'
+  kodaUrl.searchParams.set('endpoint', url)
+
+  return kodaUrl.toString()
+}
+
+export const getMetadata = async (c: HonoInterface) => {
+  const { url } = c.req.query()
+  const key = 'v1.0.0-' + encodeEndpoint(url)
+
+  // ensure ?url=url is present
+  if (!url) {
+    return c.text('url is required', 400)
+  }
+
+  try {
+    // check on KV
+    const metadataKV = await c.env.METADATA.get(key)
+    if (metadataKV) {
+      return c.json(JSON.parse(metadataKV))
+    }
+
+    // 1. put to KV
+    // ----------------------------------------
+    const externalUrl = url.includes('kodadot.xyz') ? toNftstorage(url) : url // TODO: probably not needed
+    const data = await fetch(externalUrl)
+    const json = await data.json()
+    const content = contentFrom(json as BaseMetadata)
+    // @ts-ignore
+    const normalized = normalize(content, ipfsUrl)
+
+    const image = normalized.image
+    const imageMimeType = await getMimeType(image)
+
+    const animationUrl = normalized.animationUrl
+    let animationUrlMimeType
+    if (animationUrl) {
+      animationUrlMimeType = await getMimeType(animationUrl)
+    }
+
+    const predefinedAttributes = {
+      image,
+      imageMimeType,
+      animationUrl,
+      animationUrlMimeType,
+      // TODO: get video thumbnail once we implemented CF-Streams
+      // https://image-beta.w.kodadot.xyz/ipfs/bafkreia3j75r474kgxxmptwh5n43j5nrvn3du5l7dcfq2twh73wmagqs6m
+      thumbnail: normalized.thumbnail,
+    }
+
+    if (!image.includes('w.kodadot.xyz')) {
+      predefinedAttributes.image = toCustomKodaURL(image)
+    }
+
+    if (!animationUrl.includes('w.kodadot.xyz')) {
+      predefinedAttributes.animationUrl = toCustomKodaURL(animationUrl)
+    }
+
+    const attributes = { ...predefinedAttributes, _raw: normalized }
+
```

Comment:
What does this app do?

Reply:
Ignore most parts of this section. most of the attributes should be from `hyperdata`. extended attributes from this worker should be `mimeType` and `thumbnail (once we already have cf-streams)`